Upgrade to AGP 7

jossiwolf opened the pull request.

jossiwolf committed:
Migrate to AGP 7 and update to Gradle 6.8.2

This is needed to get started with Compose

jossiwolf committed:
Update to Kotlin 1.4.30

jossiwolf committed:
Add Jetpack Compose dependencies

jossiwolf committed:
Add Auth-Compose Module as base and sample for all other Compose-related development

thatfiredev reviewed (commented):
This look good, but I'm wondering why our Android CI GitHub Action didn't run here 🤔

jossiwolf commented:
Thanks! I think I hadn't enabled GitHub Actions yet but I think I can dispatch it manually for this PR!

jossiwolf committed:
#2 Compose Infra + Auth Compose Module

jossiwolf committed:
Update Android CI to Java 11

jossiwolf committed:
Disable selected lint checks after updating to AGP 7

Added comments to most of them - let's figure these out later. A bunch is going to be done with the work needed for Firebase-UI 8.0.0.

jossiwolf committed:
Use minSDK 21 for auth-compose

jossiwolf committed:
Explicitly declare BOM dependencies in androidTestImplementation

Since AGP 4.1, handling of dependency graphs has changed a bit. That means that declaring version constraints through e.g. a BOM in the main dependency graph (`implementation`) doesn't apply them to the `androidTestImplementation` dependency graph, so we need to explicitly depend on the BOM.

We should revert this as soon as we can as this also means that any libraries depending on us through a BOM will have to declare a dependency on it in their android test graph if they have instrumented tests with Firebase dependencies.

See b/182167029 for an issue to track.

jossiwolf commented:
@rosariopfernandes I "fixed" the lint issues - I don't think this PR is the place to fix these. Let's report them to Sam so they can work on this upstream.

The latest commit will fail because of https://issuetracker.google.com/issues/182468430 / https://github.com/robolectric/robolectric/issues/6277. Do we want to merge this PR or wait until we have more info?

thatfiredev commented:
@jossiwolf I'm impressed by the number of errors that AGP 7 is giving us 😅 

The issues you linked above mention that no error is shown in `alpha08`. Do you think we can downgrade while we wait for a fix?

jossiwolf commented:
Right? Wouldn't have expected this much resistance...😂

We coooould, yes, but we'd have to downgrade to AS Canary 8 then too :/ which for me is more painful than waiting a few more days to get started😬😄

thatfiredev commented:
@jossiwolf Aahh, I forgot the AGP version is always linked to the AS version! 🤕 

Sure, let's wait a bit till AGP 7 is fixed 😓

jossiwolf commented:
I might just get started by branching off this, you could do the same - it's only tests that are failing!

jossiwolf committed:
Bump to AGP 7.0.0 Alpha 11 and Compose Beta 03

jossiwolf committed:
Update to AGP 7.0.0-alpha12

This will fix our Robolectric test issues!!

jossiwolf commented:
Ayyy, alpha 12 is out! This should fix our issues and if you're happy we can merge once CI is done :)

jossiwolf committed:
Fix Lint Issues

thatfiredev committed:
test: fix firestore data source tests

thatfiredev committed:
refactor: removed unused ids

thatfiredev committed:
refactor(firestore): suppress NotifyDataSetChanged lint error

thatfiredev committed:
chore(lint, internal): add junit as a testImplementation

thatfiredev commented on firestore/src/androidTest/java/com/firebase/ui/firestore/FirestoreDataSourceTest.java:
```diff
@@ -159,8 +159,8 @@ public void testLoadAfter_retry() throws Exception {
     private void initMockQuery() {
-        when(mMockQuery.startAfter(any())).thenReturn(mMockQuery);
-        when(mMockQuery.endBefore(any())).thenReturn(mMockQuery);
+        when(mMockQuery.startAfter(any(DocumentSnapshot.class))).thenReturn(mMockQuery);
+        when(mMockQuery.endBefore(any(DocumentSnapshot.class))).thenReturn(mMockQuery);
```
This fix should probably go to the upstream repo. I'll see if I can put up a  PR

thatfiredev commented on build.gradle.kts:
```diff
@@ -101,7 +101,16 @@ fun Project.configureAndroid() {
-                    "SelectableText", "SyntheticAccessor" // We almost never care about this
+                    "SelectableText", "SyntheticAccessor", // We almost never care about this
+                    "MediaCapabilities", // TODO: 13.03.21 Figure this one out. No clue where it comes from
+                    "NonConstantResourceId", // We're stuck with ButterKnife so we need this...
```
I think I can also remove ButterKnife from upstream so that we can remove this.

jossiwolf replied:
That would be absolutely lovely! I'm guessing Sam is planning that anyways, but very good if we can accelerate that

thatfiredev commented on build.gradle.kts:
```diff
@@ -101,7 +101,16 @@ fun Project.configureAndroid() {
+                    "ConvertToWebp", // The Firebase auth logo hasn't been converted to WebP yet
+                    "RtlHardcoded", "Bidirectional Text", "Internationalization", // There are a few lint issues we need to fix upstream
+                    "IntentFilterExportedReceiver",
+                    "UnusedResources",
```
It's curious that even though you disabled `UnusedResources` here, gradle was still reporting it as a lint error in `auth`. I tried disabling it in the `auth` module configuration directly, but the lint error was still being thrown in gradle builds.

Since we're using alpha builds here, maybe we should expect it to be broken?

jossiwolf replied:
I don't really think so... I'm guessing this is because of our module setup but I don't have a clear picture of this. I'll try to ask some AGP folks about this, the amount of Lint stuff that's broken is a bit odd

thatfiredev reviewed (approved):
@jossiwolf I managed to fix the remaining lint errors and your changes LGTM! :partying_face: 

I left a few comments, but they shouldn't stop us from merging this PR. We can address them later.
